### ISSUE TITLE
webcomponentsjs is not compatible with third-party scripts that use window.getComputedStyle

### ISSUE THREAD

**Issue by robdodson** [webcomponents/webcomponentsjs#89]
_Migrating @keanulee's original ticket from platform-dev repo:_

I was trying to include Disqus on my blog that uses platform.js (https://blog.keanulee.com/2014/10/11/setting-up-ssl-on-github-pages.html), but the Disqus script would not work when platform.js was imported. On Chrome it works, but on browsers where window.getComputedStyle is shimmed (Safari, Firefox), the Disqus script would throw an error.

Safari: TypeError: null is not an object (evaluating 't.defaultView.getComputedStyle(b,null).getPropertyValue') (getComputedStyle returns null when arg is not an element)
Firefox: TypeError: Argument 1 of Window.getComputedStyle does not implement interface Element.

I confirmed that removing platform.js fixes the Disqus script.

---

This error was caused by something in the Disqus code, but platform.js might be partly responsible. Below is a summarized version of the offending code:

```
getScrollContainer: function() {
  var elem = document.querySelector('#disqus_thread'); // some <div>
  do {
  var overflowY = getComputedStyle(div).getPropertyValue('overflow-y');
  if ("scroll" === overflowY || "auto" === overflowY) break;
  elem = elem.parentNode;
} while (elem && elem !== document.documentElement);
```

Recall that elem is a wrapped element, so when elem is the wrapped element (as it moves up the tree), it will not pass the elem !== document.documentElement

**Comment by Mikanoshi** [comment]
So is there a way to use current Disqus with Polymer? Tried putting http://a.disquscdn.com/embed.js code to IIFE, but it's not working...

**Comment by triune** [comment]
I'm experiencing a similar issue when using's ReCaptcha with Web Components. There, the same error message is coming up when I try to load the challenge questions from ReCaptcha only in Firefox and IE. However, since the ReCaptcha javascript is loaded externally, I cannot wrap() its code to get around this problem. Any suggestions here? Posted on SO as well http://stackoverflow.com/questions/28619821/recaptcha-doesnt-work-when-used-in-conjunction-with-polymer-in-firefox-and-ie

**Comment by Mikanoshi** [comment]
Download the code and wrap it :)
https://www.gstatic.com/recaptcha/api2/r20150202092117/recaptcha__en.js
But for me it's working in FF without errors...

**Comment by triune** [comment]
@Mikanoshi your suggestion does solve my problem, but it creates another in that I have to manage recaptcha code on my own. The issue was confirmed here https://github.com/GoogleWebComponents/google-recaptcha/issues/5#issuecomment-75374483 so I'd assume an upstream fix is inevitable.

**Comment by cbalit** [comment]
@triune i'n not sure to see how to deal with it, since it cames from webcomponents.js and the only available solution is to wrap recaptcha.js file. By the way you sais you made it work by wraping the recaptha api. Even that i didn't suceed. Do you have an exemple of that ?

**Comment by nazar-pc** [comment]
Setting `experiment.enable_scroll_container` option to `false` helps as of today:
```javascript
window.disqus_config = function () {
...
  this.experiment.enable_scroll_container = false;
...
};
```

**Comment by BYK** [comment]
Hey, Disqus Engineer here. Please do not download our code and then wrap it since we update it regularly :)

I'd like to help resolving this but I do not want to use a custom `wrap` method in our generic script. Do you think we can use an `instanceof` check here to know we're at the top of the document?

**Comment by BYK** [comment]
Also @nazar-pc, good find there but that would disable a very useful feature so I'd recommend against it.

**Comment by Mikanoshi** [comment]
> Please do not download our code and then wrap it since we update it regularly :)

We download and wrap it regularly too =)
```
$cache_file = "disqus.embed.js";
$url = "https://a.disquscdn.com/embed.js";
if (file_exists($cache_file) && (filemtime($cache_file) > (time() - 6 * 60 * 60))) {
	readfile($cache_file);
} else {
	$file = "(function(document) {\n".curlget($url)."\n})(wrap(document));";
	file_put_contents($cache_file, $file, LOCK_EX);
	echo $file;
}
```

But since I updated to Polymer 1.x (which uses webcomponents-lite) Disqus code is working perfectly without any modifications.

**Comment by BYK** [comment]
@Mikanoshi - we are sending different versions of the file based on the forum so what you are doing is wrong :) You should fetch it from `https://your_forum_shortname.disqus.com/embed.js`.

That said good to know it is working now!

**Comment by nazar-pc** [comment]
@BYK, you can reproduce this issue by simply including full Shadow DOM polyfill on any page with Disqus widget.
As soon as you can reproduce an issue, it shouldn't take too much to figure out what is going on (few of us already debugged minified sources and kind of fixed it on our side).

I believe you'll be able to fix it in some generic way instead of needing to call `wrap()` (I personally didn't have any single case where it was not possible to do something in generic way during last 2 years or so). If you need some more help specific to Web Components or polyfill - feel free to ask here or even reach me directly.

**Comment by BYK** [comment]
Okay, I've spent some time on this and it looks like the polyfill is leaky. I don't think expecting third-party scripts to be aware of a polyfill and then try to use methods provided that is a reasonable expectation. Do you see any other way that we can fix this? Right now the only way seems to be checking `window.ShadowDOMPolyfill` and using `ShadowDOMPolyfill.unwrap` wherever we need to check if something is the `document` or `document.documentElement` or wrap any places we protect for these with `try`/`catch` blocks which is more acceptable but still extra maintenance and a performance hit.

**Comment by nazar-pc** [comment]
What do you think about `document.querySelector('html')`?
It will return wrapped `document.documentElement` you need.

**Comment by BYK** [comment]
@nazar-pc that still doesn't solve the direct `document` comparison issue. I think I got it working with a localized try-catch block so I'll try to go with that. That said I still think this is something the polyfill should fix itself.

**Comment by nazar-pc** [comment]
Can you post specific example code where you have comparison?

> That said I still think this is something the polyfill should fix itself.

Ideally - yes, but practically virtually impossible.

**Comment by nazar-pc** [comment]
```javascript
document.querySelector('html').parentNode
```
Will give you wrapped `document` in universal way :smile:

**Comment by BYK** [comment]
@nazar-pc we've have at least two places where we either do `node === document` or `elm === document.documentElement` (not sharing code since this is the crucial part) where we do `node.parentNode` in a `while` loop.

Your suggestions are pretty interesting to be honest but I don't wanna spread these over the code base if I can fix this with a centralized and focused try-catch block. Will update this thread once the patch is out on production :)

**Comment by nazar-pc** [comment]
If you have centralized place - you can also use there wrapped `document` and wrapped `document.documentElement` using approach I've suggested. Try-catch seems to be even more hacky than this, since it captures error, while it is possible gracefully avoid it.

Anyway, I do not care about particular solution on your side if it works fine:)

**Comment by nazar-pc** [comment]
@BYK, do you have any estimation when an update will be released?

**Comment by BYK** [comment]
@nazar-pc I was hoping to get the fix out this week but other things came up. Basically, the code is ready and awaiting for review. We had to do 2 iterations since fixing the loading issue revealed some other minor issues related to the same root issue (`documentElement` being wrapped and not passing the equality checks from `document.documentElement`). I had to use your `document.querySelector('html')` trick which seems to help so thanks for that ;)

TL;DR - The fix should be out early next week.
